src/category_theory/action.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2020 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+
+import category_theory.groupoid
+import category_theory.endomorphism
+import category_theory.elements
+import group_theory.group_action
+import category_theory.single_obj
+
+/-!
+# Actions as functors and as categories
+
+From a multiplicative action M ↻ X, we can construct a functor from M to the category of
+types, mapping the single object of M to X and an element `m : M` to map `X → X` given by
+multiplication by `m`.
+  This functor induces a category structure on X -- a special case of the category of elements.
+A morphism `x → y` in this category is simply a scalar `m : M` such that `m • x = y`. In the case
+where M is a group, this category is a groupoid -- the `action groupoid'.
+-/
+
+open mul_action
+namespace category_theory
+
+universes u
+
+variables (M : Type*) [monoid M] (X : Type u) [𝒜 : mul_action M X]
+include 𝒜
+
+/-- A multiplicative action M ↻ X viewed as a functor mapping the single object of M to X
+  and an element `m : M` to the map `X → X` given by multiplication by `m`. -/
+@[reducible]
+def action_as_functor : single_obj M ⥤ Type u :=
```

Comment:
```suggestion
@[simps]
def action_as_functor : single_obj M ⥤ Type u :=
```
I would prefer `simps` over `reducible`. You never want the entire `action_as_functor` spilling its guts into the goal, just the projections to simplify.

Reply:
I think we want `p : action_category M X`, `p.2 : (action_as_functor M X).obj p.1` to reduce to `p.2 : X`. Replacing `reducible` with `simps` breaks `hom_as_subtype`; the statement doesn't type check. Is there an attribute that would ensure `(action_as_functor M X).obj _` always reduces to `X`?

Reply:
Does it help if you write the simp-lemmas manually?

Reply:
It's not a case for simp I think. Actually `hom_as_subtype` should have been written in terms of `obj_equiv` instead of `p.2` -- with this change, we don't need the `reducible` attribute.